pandas/core/frame.py

```diff
@@ -2727,11 +2727,16 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
                   verify_integrity=False):
         """
         Set the DataFrame index (row labels) using one or more existing
-        columns. By default yields a new object.
+        columns and/or new arrays of values. By default yields a new object.
 
         Parameters
         ----------
-        keys : column label or list of column labels / arrays
+        keys : column label (str), Index, Series, array, or a list of these things
+            Existing columns to set as the index (when given columns labels)
+            and/or new values to set as the index. If an Index is given, it's
```

Comment:
This gives an ambiguous case where the list-like has all of the same column names. So I would rather have an unambiguous definition. Require ALL of the labels to be present in the existing index (then just use them). A ValueError would be raised if the len doesn't match in any event (e.g. with a passed list-like).